Mathlib/CategoryTheory/Comma/Over.lean

```diff
@@ -657,4 +657,114 @@ lemma toUnder_comp_forget (F : S ⥤ T) (X : T) (f : (Y : S) → X ⟶ F.obj Y)
+      (fun f => f.right.hom) (by simp))
+    (Functor.toStructuredArrow
+      (Functor.toStructuredArrow (StructuredArrow.proj Y _ ⋙ Under.forget X) _ _
+         (fun g => by exact g.hom) (fun m => by have := m.w; aesop_cat)) _ _
+      (fun f => f.right.hom) (by simp))
+    (NatIso.ofComponents (fun x => by
+      simp only [Functor.id_obj, Functor.const_obj_obj, proj_obj, left_eq_id, Functor.const_obj_map,
+        proj_map, Functor.comp_obj, Functor.comp_map, eq_mp_eq_cast, id_eq, Under.forget_obj,
+        Under.forget_map, Functor.toStructuredArrow_obj, Functor.toUnder_obj, mk_right,
+        mk_hom_eq_self]
+      apply eta) (by simp))
+    (NatIso.ofComponents (fun x => by
+      simp only [Functor.const_obj_obj, Functor.comp_obj, Under.forget_obj, left_eq_id,
+        Functor.const_obj_map, Functor.comp_map, Under.forget_map, proj_obj, proj_map,
+        eq_mp_eq_cast, id_eq, Functor.toStructuredArrow_obj, Functor.toUnder_obj, mk_right,
+        mk_hom_eq_self, Functor.id_obj]
+      symm
+      apply eta) (by aesop_cat))
+
+/-- Characterization of the structured arrow category on the diagonal functor `T ⥤ T × T`. -/
+def ofDiagEquivalence (X : T × T) :
+    StructuredArrow X (Functor.diag _) ≌ StructuredArrow X.2 (Under.forget X.1) :=
+  Equivalence.mk
+    (Functor.toStructuredArrow
+      (Functor.toUnder (StructuredArrow.proj X _) _
+        (fun f => by exact f.hom.1) (fun m => by have := m.w; aesop_cat)) _ _
+      (fun f => f.hom.2) (fun m => by have := m.w; aesop_cat))
+    (Functor.toStructuredArrow (StructuredArrow.proj _ _ ⋙ Under.forget _) _ _
+      (fun f => (f.right.hom, f.hom)) (fun m => by have := m.w; aesop_cat))
+    (NatIso.ofComponents (fun x => by
+      simp only [Functor.id_obj, Functor.const_obj_obj, Functor.diag_obj, prod_Hom, left_eq_id,
+        Functor.const_obj_map, prod_comp, prod_id_fst, prod_id_snd, Functor.diag_map, proj_obj,
+        proj_map, eq_mp_eq_cast, id_eq, Functor.comp_obj, Under.forget_obj,
+        Functor.toStructuredArrow_obj, Functor.toUnder_obj, mk_right, mk_hom_eq_self, Prod.mk.eta]
+      apply eta) (by simp))
+    (NatIso.ofComponents (fun x => by
+      simp only [Functor.comp_obj, proj_obj, Under.forget_obj, Functor.diag_obj,
+        Functor.const_obj_obj, prod_Hom, left_eq_id, Functor.const_obj_map, prod_comp,
+        Functor.id_obj, prod_id_fst, prod_id_snd, Functor.diag_map, proj_map, eq_mp_eq_cast, id_eq,
+        Functor.toStructuredArrow_obj, Functor.toUnder_obj, mk_right, mk_hom_eq_self]
+      symm
+      apply eta) (by aesop_cat))
+
+end StructuredArrow
+
+namespace CostructuredArrow
+
+/-- Characterization of the costructured arrow category on the projection functor of any
+costructured arrow category. -/
+def ofCostructuredArrowProjEquivalence {D : Type u₂} [Category.{v₂} D]
+    (F : T ⥤ D) (Y : D) (X : T) :
+    CostructuredArrow (CostructuredArrow.proj F Y) X ≌ CostructuredArrow (Over.forget X ⋙ F) Y :=
```

Comment:
It would be slightly more readable to use the following syntax:
```suggestion
    CostructuredArrow (CostructuredArrow.proj F Y) X ≌ CostructuredArrow (Over.forget X ⋙ F) Y where
  functor := sorry
  inverse := sorry
  unitIso := sorry
  counitIso := sorry
```

Reply:
I would say that using `Equivalence.mk` is acceptable because it is a "smart" constructor that changes the unit and counit so that the triangle equality is automatically satisfied.

Reply:
Yes, that was a conscious choice, will try if proving the triangle equality is cheap, otherwise I'd keep it this way.

Reply:
All the triangle equalities are proved automatically (and the file would take a very similar amount of time to compile on my computer). In https://github.com/leanprover-community/mathlib4/pull/15919 I removed almost all the unnecessary uses of `Equivalence.mk` (and surprisingly, it made mathlib faster! I do not know how...). I do not think we should speculate about whether nobody would ever need to unfold the definitions of the (co)unit isomorphisms.